tests/flow.ml

```diff
@@ -0,0 +1,219 @@
+  let equal a b = match a, b with
+    | `Msg a, `Msg b -> a = b
+    | `Not_found, `Not_found -> true
+    | _ -> false in
+  Alcotest.testable pp equal
+
+module Memory_flow0 = struct
+  type input = bytes and output = string
+  type +'a s = 'a
+
+  type flow =
+    { mutable i : string
+    ; o : bytes
+    ; mutable p : int
+    ; mutable c : bool }
+
+  type endpoint = (string * bytes)
+  type error = [ `Closed ]
+
+  let closed_by_peer = "Closed by peer"
+  let pp_error ppf = function
+    | `Closed -> Fmt.string ppf closed_by_peer
+
+  let connect (i, o) = Ok { i; o; p= 0; c= false; }
+
+  let recv flow buf =
+    let len = min (String.length flow.i) (Bytes.length buf) in
+    if len = 0 then ( flow.c <- true ; Ok `End_of_flow )
+    else
+      ( Bytes.blit_string flow.i 0 buf 0 len
+      ; flow.i <- String.sub flow.i len (String.length flow.i - len)
+      ; Ok (`Input len) )
+
+  let send flow str =
+    if flow.c then Error `Closed
+    else
+      ( let len = min (Bytes.length flow.o - flow.p) (String.length str) in
+        Bytes.blit_string str 0 flow.o flow.p len
+      ; flow.p <- flow.p + len
+      ; Ok len )
+
+  let close flow = flow.c <- true ; Ok ()
+end
+
+let memory0 = Conduit.register ~protocol:(module Memory_flow0)
+
+let test_input_string =
+  Alcotest.test_case "input string" `Quick @@ fun () ->
+  let open Rresult in
+  let flow = Conduit.connect ("Hello World!", Bytes.empty) memory0 in
+  Alcotest.(check bool) "connect" (R.is_ok flow) true ;
+  let flow = R.get_ok flow in
+  let buf0 = Bytes.create 12 in
+  let buf1 = Bytes.create 12 in
+  let res0 = Conduit.recv flow buf0 in
+  let res1 = Conduit.recv flow buf1 in
+  let res2 = Conduit.send flow "Hello World!" in
+  Alcotest.(check (result recv error)) "res0" res0 (Ok (`Input 12)) ;
+  Alcotest.(check string) "buf0" (Bytes.to_string buf0) "Hello World!" ;
+  Alcotest.(check (result recv error)) "res1" res1 (Ok `End_of_flow) ;
+  Alcotest.(check (result send error)) "res2" res2 (Error (`Msg Memory_flow0.closed_by_peer)) ;
+;;
+
+let test_output_string =
+  Alcotest.test_case "output string" `Quick @@ fun () ->
+  let open Rresult in
+  let buf = Bytes.create 12 in
+  let flow = Conduit.connect ("", buf) memory0 in
+  Alcotest.(check bool) "connect" (R.is_ok flow) true ;
+  let flow = R.get_ok flow in
+  let res0 = Conduit.send flow "Hell" in
+  let res1 = Conduit.send flow "o Wo" in
+  let res2 = Conduit.send flow "rld!" in
+  let res3 = Conduit.send flow "?!?!" in
+  let res4 = Conduit.recv flow Bytes.empty in
+  Alcotest.(check (result send error)) "res0" res0 (Ok 4) ;
+  Alcotest.(check (result send error)) "res1" res1 (Ok 4) ;
+  Alcotest.(check (result send error)) "res2" res2 (Ok 4) ;
+  Alcotest.(check (result send error)) "res3" res3 (Ok 0) ;
+  Alcotest.(check (result recv error)) "res4" res4 (Ok `End_of_flow) ;
+  Alcotest.(check string) "buf" (Bytes.to_string buf) "Hello World!" ;
+;;
+
+module Memory_flow1 = struct
```

Comment:
What's the difference with `Memory_flow0` ?

Reply:
`Memory_flow0` computes a `string` as an input and a `bytes` as an output.
`Memory_flow1` computes a `string list` as an input and a `bytes list` as an output.

The second can be the first if we `List.concat`.